t/50reverse-proxy-early-response.t

```diff
@@ -421,12 +422,17 @@ sub read {
     $buf;
 }
 
+# This code is originated from socketpool.c
+# Find for the comment "test if the connection is still alive".
 sub is_alive {
     my ($self) = @_;
     return undef unless $self->{sock};
     my $buf;
     my $ret = $self->{sock}->recv($buf, 1, MSG_PEEK);
-    return $ret || ($! == EAGAIN || $! == EWOULDBLOCK);
+    return (
+        (defined($ret) && length($buf) > 0)
+        || ($! == EAGAIN || $! == EWOULDBLOCK)
```

Comment:
Let's say that the socket has been closed by the peer (using FIN). Then, wouldn't the output be `$ret` == undef and `$buf` == 0, with `$!` _not_ being updated? At least that is how recv (2) behaves. When the peer closes the socket, the return value of recv (2) is zero, and because that is _not_ an error, errno is not updated.

But in the current form of this PR, we are consulting `$!`.

Reply:
It makes sense to me, but I wonder why I didn't run into this problem under debugging this test, reading perl source code, and I've found this:

https://github.com/Perl/perl5/blob/blead/pp_sys.c#L1711

It means that *some* Perl's syscall warppers, including  `recv` (covered by `pp_sysread`), clears `errno` before invoking syscall. To make matters worse, it's not documented at least in `perldoc -f sysread` nor `perldoc -f recv`.

Reply:
Oops. Misunderstood. It's not related to the Perl's behavior.

Fixed: https://github.com/h2o/h2o/pull/2506/commits/c83d472b3ba8f0dbdf66531c28cdb325738e80fe